rs/replicated_state/best-practices-panics.md

```diff
@@ -0,0 +1,59 @@
+
+- A system resource is exhausted. E.g. we do not have enough file descriptors and cannot create a new file in the State Manager. There’s not much we can do about it automatically, so we should panic in that case and handle it offline.
+- An invariant in the implementation of a data structure is broken. In these cases, the data structure can perform any invariant checks internally in its implementation which is typically easier to verify/inspect if everything is contained within the data structure. An example of this is the [TaskQueue](https://sourcegraph.com/github.com/dfinity/ic@d7cac19658a397f862f9e162c32ac02d21a3d77d/-/blob/rs/replicated_state/src/canister_state/system_state.rs?L284) data structure. It’s also a good idea to have tests that cover any panics added and these should be typically safe to use. Alternatively, you should consider if debug_asserts would potentially be enough assuming we have good test coverage for the data structure.
+- An invariant is broken which could lead to corrupted state or state divergence or some otherwise broken state that is difficult to recover from. In that case a panic should be preferred to avoid creating further more difficult to resolve issues. A good example of this would be the case where we handle [reserved cycles](https://sourcegraph.com/github.com/dfinity/ic@d7cac19658a397f862f9e162c32ac02d21a3d77d/-/blob/rs/system_api/src/sandbox_safe_system_state.rs?L514) in the system API.
+- Hard invariants as defined above can result in panics when they get violated.
+
+
+## When should panics be avoided
+
+In some cases it’s best to avoid panics and instead either try to handle the errors or raise critical errors instead if it’s not easy to handle the error in place. debug_asserts are also highly encouraged for these cases.
+
+- Wrong user input should never result in a replica panic.
+- Invariants across components are very hard to maintain and keep up to date as code evolves. Avoid adding panics related to such invariants and instead use critical errors instead to make the teams aware about instances where we might hit an unexpected case (some exceptions might apply if the invariant could lead to corrupted state or hard to recover from but it really should be the last resort). A good example here is [load_canister_snapshot](https://www.google.com/url?q=https://sourcegraph.com/github.com/dfinity/ic@d7cac19658a397f862f9e162c32ac02d21a3d77d/-/blob/rs/execution_environment/src/canister_manager.rs?L2108-2126&sa=D&source=docs&ust=1733308984480434&usg=AOvVaw2K8SR6s3VMBOkSTryuhxga) where we check whether some scheduling precondition holds in the canister manager. In this case, using a panic would be very aggressive, stopping progress of the whole subnet for a bug related to canister snapshots. Returning an error back to the user and raising a critical error is the more appropriate handling of this case, containing the damage to some snapshots functionality only not working in case such a bug exists.
```

Comment:
```suggestion
- Invariants across components are very hard to maintain and keep up to date as code evolves. Avoid adding panics related to such invariants and use critical errors instead to make the teams aware about instances where we might hit an unexpected case (some exceptions might apply if the invariant could lead to corrupted state or be hard to recover from, but it really should be the last resort). A good example here is [load_canister_snapshot](https://www.google.com/url?q=https://sourcegraph.com/github.com/dfinity/ic@d7cac19658a397f862f9e162c32ac02d21a3d77d/-/blob/rs/execution_environment/src/canister_manager.rs?L2108-2126&sa=D&source=docs&ust=1733308984480434&usg=AOvVaw2K8SR6s3VMBOkSTryuhxga) where we check whether some scheduling precondition holds in the canister manager. In this case, using a panic would be very aggressive, stopping progress of the whole subnet for a bug related to canister snapshots. Returning an error back to the user and raising a critical error is the more appropriate handling of this case, containing the damage to some snapshots functionality only not working in case such a bug exists.
```